File: examples/rust/src/value.rs
// SPDX-License-Identifier: Apache-2.0
//! Words and values: what crosses an edge, and the few things built on it.
//!
//! A [`Value`] is I-JSON held to the value rule. The type carries most of the
//! rule on its own: a [`Value::String`] is valid Unicode text, a [`Number`] is
//! a finite double that is never minus zero, and a [`Map`] holds each key
//! once. What a type cannot hold, the depth bound and the rules about JSON
//! text, is checked
//! where text is read ([`Value::parse`]) and where a value is written
//! ([`Value::check`]).

mod blueprint;
mod canonical;
mod json;
mod number;
mod words;

pub use blueprint::{digest, Blueprint};
pub use number::Number;
pub use words::Word;

use std::fmt;

/// How deep a value may nest: sixty-four containers, and a sixty-fifth is not
/// a value. A scalar alone is depth zero, `[]` is depth one.
pub const MAX_DEPTH: usize = 64;

/// How deep a payload may nest. Depth is counted from the value written and
/// never from a root it is written under, so an arg of depth sixty-four stands
/// whatever carries it. The payload is a value at its own root, one object
/// holding `args` holding the arg, so it nests no deeper than sixty-six, and
/// that bound is on the payload and never on the arg.
pub const MAX_PAYLOAD_DEPTH: usize = MAX_DEPTH + 2;

/// How deep a reply may nest: one object holding `object`, so sixty-five, and
/// that bound is on the reply and never on the object.
pub const MAX_REPLY_DEPTH: usize = MAX_DEPTH + 1;

/// One I-JSON value.
#[derive(Clone, Debug, PartialEq)]
pub enum Value {
    Null,
    Bool(bool),
    Number(Number),
    String(String),
    Array(Vec<Value>),
    Object(Map),
}

/// A JSON object: each key once, in the order it was first inserted.
///
/// The order is kept because a JSON body is written in an order a vector pins;
/// the digest never reads it, since the canonical form sorts.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct Map {
    entries: Vec<(String, Value)>,
}

/// Why something is not a value.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum ValueError {
    /// Not JSON text at all, at this byte offset.
    Syntax(usize),
    /// Bytes that are not UTF-8.
    NotUtf8,
    /// An object naming one key twice.
    DuplicateKey(String),
    /// A surrogate escape that does not stand in a pair.
    LoneSurrogate,
    /// A negative zero, `-0` or `-0.0` alike.
    MinusZero,
    /// A number no double holds: an integer a double would round, a value
    /// past the largest double, or a non-zero value that rounds to zero.
    NotDouble(String),
    /// Nested past [`MAX_DEPTH`].
    TooDeep,
    /// A root key of cells that is the ward's.
    WardKey(String),
    /// A value where the ward reads a record of another shape.
    NotObject,
}

impl fmt::Display for ValueError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ValueError::Syntax(at) => write!(f, "not JSON at byte {at}"),
            ValueError::NotUtf8 => write!(f, "not UTF-8"),
            ValueError::DuplicateKey(k) => write!(f, "duplicate key {k:?}"),
            ValueError::LoneSurrogate => write!(f, "a lone surrogate"),
            ValueError::MinusZero => write!(f, "minus zero is not a value"),
            ValueError::NotDouble(t) => write!(f, "{t} is not a double"),
            ValueError::TooDeep => f.write_str("nested past the bound"),
            ValueError::WardKey(k) => write!(f, "{k:?} is the ward's"),
            ValueError::NotObject => write!(f, "not an object"),
        }
    }
}

impl std::error::Error for ValueError {}

impl Value {
    /// Reads JSON text as a value, refusing whatever the value rule refuses.
    pub fn parse(text: &str) -> Result<Value, ValueError> {
        json::parse(text)
    }

    /// Reads the bytes of a payload, which is a value at its own root and
    /// nests no deeper than [`MAX_PAYLOAD_DEPTH`]. The args it carries are
    /// counted from themselves and held to [`MAX_DEPTH`] where they are read.
    pub fn parse_payload(bytes: &[u8]) -> Result<Value, ValueError> {
        let text = std::str::from_utf8(bytes).map_err(|_| ValueError::NotUtf8)?;
        json::parse_under(text, MAX_PAYLOAD_DEPTH)
    }

    /// Reads the bytes of a reply, which is a value at its own root and nests
    /// no deeper than [`MAX_REPLY_DEPTH`]. The object it carries is held to
    /// [`MAX_DEPTH`] where it is read.
    pub fn parse_reply(bytes: &[u8]) -> Result<Value, ValueError> {
        let text = std::str::from_utf8(bytes).map_err(|_| ValueError::NotUtf8)?;
        json::parse_under(text, MAX_REPLY_DEPTH)
    }

    /// Whether this value may be written or sent. Everything but the depth is
    /// already held by the type.
    pub fn check(&self) -> Result<(), ValueError> {
        if self.depth() > MAX_DEPTH {
            Err(ValueError::TooDeep)
        } else {
            Ok(())
        }
    }

    /// How many containers deep this value nests.
    pub fn depth(&self) -> usize {
        match self {
            Value::Array(items) => 1 + items.iter().map(Value::depth).max().unwrap_or(0),
            Value::Object(map) => 1 + map.values().map(Value::depth).max().unwrap_or(0),
            _ => 0,
        }
    }

    /// JSON text, keys in their own order, numbers as ECMAScript writes them.
    pub fn to_json(&self) -> String {
        let mut out = String::new();
        canonical::write(self, false, &mut out);
        out
    }

    /// The RFC 8785 canonical form: keys by UTF-16 code unit, ES numbers.
    pub fn canonical(&self) -> String {
        let mut out = String::new();
        canonical::write(self, true, &mut out);
        out
    }

    pub fn is_null(&self) -> bool {
        matches!(self, Value::Null)
    }

    pub fn as_str(&self) -> Option<&str> {
        match self {
            Value::String(s) => Some(s),
            _ => None,
        }
    }

    pub fn as_bool(&self) -> Option<bool> {
        match self {
            Value::Bool(b) => Some(*b),
            _ => None,
        }
    }

    pub fn as_number(&self) -> Option<Number> {
        match self {
            Value::Number(n) => Some(*n),
            _ => None,
        }
    }

    pub fn as_array(&self) -> Option<&Vec<Value>> {
        match self {
            Value::Array(a) => Some(a),
            _ => None,
        }
    }

    pub fn as_object(&self) -> Option<&Map> {
        match self {
            Value::Object(m) => Some(m),
            _ => None,
        }
    }

    pub fn as_object_mut(&mut self) -> Option<&mut Map> {
        match self {
            Value::Object(m) => Some(m),
            _ => None,
        }
    }

    /// The field `key` of an object, or `None` for a missing field and for a
    /// value that is not an object.
    pub fn get(&self, key: &str) -> Option<&Value> {
        self.as_object().and_then(|m| m.get(key))
    }
}

impl fmt::Display for Value {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.to_json())
    }
}

impl From<bool> for Value {
    fn from(b: bool) -> Value {
        Value::Bool(b)
    }
}

impl From<&str> for Value {
    fn from(s: &str) -> Value {
        Value::String(s.to_owned())
    }
}

impl From<String> for Value {
    fn from(s: String) -> Value {
        Value::String(s)
    }
}

impl From<Number> for Value {
    fn from(n: Number) -> Value {
        Value::Number(n)
    }
}

impl From<Map> for Value {
    fn from(m: Map) -> Value {
        Value::Object(m)
    }
}

impl From<Vec<Value>> for Value {
    fn from(items: Vec<Value>) -> Value {
        Value::Array(items)
    }
}

impl<T: Into<Value>> From<Option<T>> for Value {
    fn from(o: Option<T>) -> Value {
        o.map_or(Value::Null, Into::into)
    }
}

impl Map {
    pub fn new() -> Map {
        Map::default()
    }

    /// How many keys the object names.
    pub fn count(&self) -> usize {
        self.entries.len()
    }

    pub fn get(&self, key: &str) -> Option<&Value> {
        self.entries.iter().find(|(k, _)| k == key).map(|(_, v)| v)
    }

    pub fn get_mut(&mut self, key: &str) -> Option<&mut Value> {
        self.entries.iter_mut().find(|(k, _)| k == key).map(|(_, v)| v)
    }

    pub fn contains_key(&self, key: &str) -> bool {
        self.get(key).is_some()
    }

    /// Sets `key`, keeping its place if it was already there, and returns
    /// what it held.
    pub fn insert(&mut self, key: impl Into<String>, value: impl Into<Value>) -> Option<Value> {
        let key = key.into();
        let value = value.into();
        match self.get_mut(&key) {
            Some(slot) => Some(std::mem::replace(slot, value)),
            None => {
                self.entries.push((key, value));
                None
            }
        }
    }

    pub fn remove(&mut self, key: &str) -> Option<Value> {
        let at = self.entries.iter().position(|(k, _)| k == key)?;
        Some(self.entries.remove(at).1)
    }

    pub fn iter(&self) -> impl Iterator<Item = (&str, &Value)> {
        self.entries.iter().map(|(k, v)| (k.as_str(), v))
    }

    pub fn values(&self) -> impl Iterator<Item = &Value> {
        self.entries.iter().map(|(_, v)| v)
    }
}

impl<K: Into<String>, V: Into<Value>> FromIterator<(K, V)> for Map {
    fn from_iter<I: IntoIterator<Item = (K, V)>>(iter: I) -> Map {
        let mut map = Map::new();
        for (k, v) in iter {
            map.insert(k, v);
        }
        map
    }
}

File: examples/go/value.go
package main

import (
	"errors"
	"strconv"
	"strings"
	"unicode/utf16"
	"unicode/utf8"
)

// A parsed value. null is nil, a boolean is bool, a string is string, a
// number is *Num. A text whose outermost value is an object is *Obj. A
// container inside that object, or an outermost array, is Container: its
// grammar is checked and nothing inside it is read.
type Num struct {
	Text string // the number as written
}

type Obj struct {
	Keys []string          // in the order written
	M    map[string]any    // parsed
	Raw  map[string][]byte // each value's bytes exactly as they arrived
}

type Container struct {
	Object bool
}

var errNotValue = errors.New("not a value")

// Parse reads JSON text. Only the outermost object's own keys are held to
// one name each. Inside it any JSON text is a value, whatever its nesting,
// strings or keys.
func Parse(b []byte) (any, error) {
	if !utf8.Valid(b) {
		return nil, errNotValue
	}
	p := &parser{s: b}
	p.ws()
	var v any
	var err error
	if p.at('{') {
		v, err = p.object()
	} else {
		v, err = p.value()
	}
	if err != nil {
		return nil, err
	}
	p.ws()
	if p.i != len(p.s) {
		return nil, errNotValue
	}
	return v, nil
}

type parser struct {
	s []byte
	i int
}

// ws skips the four whitespace bytes of RFC 8259 and nothing else.
func (p *parser) ws() {
	for p.i < len(p.s) {
		switch p.s[p.i] {
		case ' ', '\t', '\n', '\r':
			p.i++
		default:
			return
		}
	}
}

func (p *parser) at(c byte) bool { return p.i < len(p.s) && p.s[p.i] == c }

func (p *parser) lit(word string) bool {
	if strings.HasPrefix(string(p.s[p.i:min(len(p.s), p.i+len(word))]), word) {
		p.i += len(word)
		return true
	}
	return false
}

// value reads one value. A container is skipped over and answered as its kind.
func (p *parser) value() (any, error) {
	if p.i >= len(p.s) {
		return nil, errNotValue
	}
	switch c := p.s[p.i]; {
	case c == '{' || c == '[':
		return Container{Object: c == '{'}, p.skip()
	case c == '"':
		return p.str()
	case c == 't':
		if p.lit("true") {
			return true, nil
		}
	case c == 'f':
		if p.lit("false") {
			return false, nil
		}
	case c == 'n':
		if p.lit("null") {
			return nil, nil
		}
	case c == '-' || (c >= '0' && c <= '9'):
		return p.number()
	}
	return nil, errNotValue
}

// key reads an object key and its colon.
func (p *parser) key() (string, error) {
	p.ws()
	if !p.at('"') {
		return "", errNotValue
	}
	k, err := p.str()
	if err != nil {
		return "", err
	}
	p.ws()
	if !p.at(':') {
		return "", errNotValue
	}
	p.i++
	p.ws()
	return k.(string), nil
}

// skip checks a container against the grammar without recursion, so no
// nesting is too deep to read.
func (p *parser) skip() error {
	var stack []byte
	for {
		// a value is owed here
		p.ws()
		if p.at('{') || p.at('[') {
			c := p.s[p.i]
			p.i++
			p.ws()
			if p.at(c + 2) { // '}' is '{'+2, ']' is '['+2
				p.i++
			} else {
				stack = append(stack, c)
				if c == '{' {
					if _, err := p.key(); err != nil {
						return err
					}
				}
				continue
			}
		} else if _, err := p.value(); err != nil {
			return err
		}
		// a value was read
		for {
			if len(stack) == 0 {
				return nil
			}
			p.ws()
			top := stack[len(stack)-1]
			if p.at(',') {
				p.i++
				if top == '{' {
					if _, err := p.key(); err != nil {
						return err
					}
				}
				break
			}
			if !p.at(top + 2) {
				return errNotValue
			}
			p.i++
			stack = stack[:len(stack)-1]
		}
	}
}

// object reads the outermost object, whose own keys are one name each.
func (p *parser) object() (any, error) {
	p.i++
	o := &Obj{M: map[string]any{}, Raw: map[string][]byte{}}
	p.ws()
	if p.at('}') {
		p.i++
		return o, nil
	}
	for {
		key, err := p.key()
		if err != nil {
			return nil, err
		}
		// two keys are one name when equal after escapes are read
		if _, dup := o.M[key]; dup {
			return nil, errNotValue
		}
		start := p.i
		v, err := p.value()
		if err != nil {
			return nil, err
		}
		o.Keys = append(o.Keys, key)
		o.M[key] = v
		o.Raw[key] = p.s[start:p.i]
		p.ws()
		switch {
		case p.at(','):
			p.i++
		case p.at('}'):
			p.i++
			return o, nil
		default:
			return nil, errNotValue
		}
	}
}

func hex4(b []byte) (rune, bool) {
	if len(b) < 4 {
		return 0, false
	}
	for _, c := range b[:4] {
		if !(c >= '0' && c <= '9' || c >= 'a' && c <= 'f' || c >= 'A' && c <= 'F') {
			return 0, false
		}
	}
	v, err := strconv.ParseUint(string(b[:4]), 16, 16)
	return rune(v), err == nil
}

// str reads a string. A lone surrogate is a value, and is kept as its
// generalized UTF-8 bytes, which no UTF-8 text spells, so two strings are
// equal exactly when their code units are.
func (p *parser) str() (any, error) {
	p.i++ // the opening quote
	var sb strings.Builder
	for {
		if p.i >= len(p.s) {
			return nil, errNotValue
		}
		c := p.s[p.i]
		switch {
		case c == '"':
			p.i++
			return sb.String(), nil
		case c < 0x20:
			return nil, errNotValue
		case c == '\\':
			if p.i+1 >= len(p.s) {
				return nil, errNotValue
			}
			e := p.s[p.i+1]
			p.i += 2
			switch e {
			case '"', '\\', '/':
				sb.WriteByte(e)
			case 'b':
				sb.WriteByte('\b')
			case 'f':
				sb.WriteByte('\f')
			case 'n':
				sb.WriteByte('\n')
			case 'r':
				sb.WriteByte('\r')
			case 't':
				sb.WriteByte('\t')
			case 'u':
				r, ok := hex4(p.s[p.i:])
				if !ok {
					return nil, errNotValue
				}
				p.i += 4
				if r >= 0xd800 && r < 0xdc00 && p.i+6 <= len(p.s) && p.s[p.i] == '\\' && p.s[p.i+1] == 'u' {
					if r2, ok := hex4(p.s[p.i+2:]); ok && r2 >= 0xdc00 && r2 <= 0xdfff {
						p.i += 6
						r = utf16.DecodeRune(r, r2)
					}
				}
				if utf16.IsSurrogate(r) {
					sb.Write([]byte{byte(0xe0 | r>>12), byte(0x80 | (r>>6)&0x3f), byte(0x80 | r&0x3f)})
				} else {
					sb.WriteRune(r)
				}
			default:
				return nil, errNotValue
			}
		default:
			sb.WriteByte(c)
			p.i++
		}
	}
}

// number reads any number the grammar of RFC 8259 allows, as written.
func (p *parser) number() (any, error) {
	st := p.i
	if p.at('-') {
		p.i++
	}
	digits := func() int {
		n := 0
		for p.i < len(p.s) && p.s[p.i] >= '0' && p.s[p.i] <= '9' {
			p.i++
			n++
		}
		return n
	}
	if p.at('0') {
		p.i++
	} else if digits() == 0 {
		return nil, errNotValue
	}
	if p.at('.') {
		p.i++
		if digits() == 0 {
			return nil, errNotValue
		}
	}
	if p.at('e') || p.at('E') {
		p.i++
		if p.at('+') || p.at('-') {
			p.i++
		}
		if digits() == 0 {
			return nil, errNotValue
		}
	}
	return &Num{Text: string(p.s[st:p.i])}, nil
}

// MaxCount is 2^53 - 1, the highest count number.
const MaxCount = 1<<53 - 1

// CountOf answers a count number: no fraction, no exponent, 1 to 2^53 - 1.
// The grammar already refused a leading zero.
func CountOf(v any) (uint64, bool) {
	n, ok := v.(*Num)
	if !ok || strings.ContainsAny(n.Text, "-.eE") {
		return 0, false
	}
	c, err := strconv.ParseUint(n.Text, 10, 64)
	return c, err == nil && c >= 1 && c <= MaxCount
}

// Quote writes a string as JSON text.
func Quote(s string) []byte {
	var sb strings.Builder
	sb.WriteByte('"')
	for _, r := range s {
		switch {
		case r == '"':
			sb.WriteString(`\"`)
		case r == '\\':
			sb.WriteString(`\\`)
		case r < 0x20:
			sb.WriteString(`\u00`)
			sb.WriteByte("0123456789abcdef"[r>>4])
			sb.WriteByte("0123456789abcdef"[r&15])
		default:
			sb.WriteRune(r)
		}
	}
	sb.WriteByte('"')
	return []byte(sb.String())
}
